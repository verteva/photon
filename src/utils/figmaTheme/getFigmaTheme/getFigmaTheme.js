import merge from 'lodash.merge';

// Function that sorts one array based on the sort of the other array
const sortArray = (arr, sortArr) => {
  return arr.sort((a, b) => {
    return (
      sortArr.indexOf(a.replace('.json', '')) -
      sortArr.indexOf(b.replace('.json', ''))
    );
  });
};

export const getFigmaTheme = (files, theme) => {
  const filePaths = files.keys();

  const metaFilePath = filePaths.find((filePath) =>
    filePath.includes('$metadata')
  );

  if (!metaFilePath) {
    throw new Error('No $meta file found');
  }

  const { tokenSetOrder } = files(metaFilePath);

  const themeFilePath = filePaths.find((filePath) =>
    filePath.includes('$theme')
  );

  if (!themeFilePath) {
    throw new Error('No $theme file found');
  }

  const themeFile = files(themeFilePath);

  const themeData = themeFile.find(({ name }) => name === theme);

  if (!themeData) {
    throw new Error(
      `No data found for theme ${theme}. This theme may not exist.`
    );
  }

  const sourcePartialsToLoad = tokenSetOrder.filter(
    (themeName) =>
      Boolean(themeData.selectedTokenSets[themeName]) &&
      themeData.selectedTokenSets[themeName] !== 'disabled'
  );

  const semanticPartialsToLoad = tokenSetOrder.filter(
    (themeName) => themeData.selectedTokenSets[themeName] === 'enabled'
  );

  if (!sourcePartialsToLoad || !sourcePartialsToLoad.length) {
    throw new Error(
      `No enabled partials found for this theme ${theme}. Check the $theme file for this theme.`
    );
  }

  const sortedSourcePartialsToLoad = sortArray(filePaths, tokenSetOrder).filter(
    (sortedFile) =>
      sourcePartialsToLoad.includes(sortedFile.replace('.json', ''))
  );

  const sortedSemanticPartialsToLoad = sortArray(
    filePaths,
    tokenSetOrder
  ).filter((sortedFile) =>
    semanticPartialsToLoad.includes(sortedFile.replace('.json', ''))
  );

  const sourceTheme = sortedSourcePartialsToLoad.reduce(
    (acc, filePath) => merge(acc, files(filePath)),
    {}
  );

  const semanticTheme = sortedSemanticPartialsToLoad.reduce(
    (acc, filePath) => merge(acc, files(filePath)),
    {}
  );

  return { sourceTheme, semanticTheme };
};

export const getFigmaThemeNames = (files) => {
  const filePaths = files.keys();

  const themeFilePath = filePaths.find((filePath) =>
    filePath.includes('$theme')
  );

  if (!themeFilePath) {
    throw new Error('No $theme file found');
  }

  const themeFile = files(themeFilePath);

  return themeFile.map(({ name }) => name);
};
