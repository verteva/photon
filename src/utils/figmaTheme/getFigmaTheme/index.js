export { getFigmaTheme, getFigmaThemeNames } from './getFigmaTheme';
