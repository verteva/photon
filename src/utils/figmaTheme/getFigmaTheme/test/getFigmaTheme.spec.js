import { getFigmaTheme, getFigmaThemeNames } from '../getFigmaTheme';
import {
  correctFiles,
  correctNestedFiles,
  brokenFilesNoTheme,
  brokenFilesNoMeta,
  brokenFilesIncorectTokens,
} from '../../mockThemeFiles/testFileObjects';

const createFileObject = (files) => {
  const keys = Object.keys(files);

  const fileObject = function (key) {
    return files[key];
  };

  fileObject.keys = function () {
    return keys;
  };

  return fileObject;
};

describe('figmaTheme/getFigmaTheme', () => {
  it('gets a figma theme from a files object', () => {
    const result = getFigmaTheme(createFileObject(correctFiles), 'theme1');
    expect(result.semanticTheme).toMatchSnapshot();
    expect(result.sourceTheme).toMatchSnapshot();
  });

  it('gets a second figma theme from a files object', () => {
    const result = getFigmaTheme(createFileObject(correctFiles), 'theme2');
    expect(result.semanticTheme).toMatchSnapshot();
    expect(result.sourceTheme).toMatchSnapshot();
  });

  it('gets a figma theme from a files object with folders', () => {
    const result = getFigmaTheme(
      createFileObject(correctNestedFiles),
      'theme1'
    );
    expect(result.semanticTheme).toMatchSnapshot();
    expect(result.sourceTheme).toMatchSnapshot();
  });

  describe('errors', () => {
    it('throws when a $theme file cannot be found', () => {
      expect(() =>
        getFigmaTheme(createFileObject(brokenFilesNoTheme), 'theme1')
      ).toThrowErrorMatchingSnapshot();
    });

    it('throws when a $meta file cannot be found', () => {
      expect(() =>
        getFigmaTheme(createFileObject(brokenFilesNoMeta), 'theme1')
      ).toThrowErrorMatchingSnapshot();
    });

    it('throws when there is no theme data for supplied theme', () => {
      expect(() =>
        getFigmaTheme(createFileObject(correctFiles), 'theme4')
      ).toThrowErrorMatchingSnapshot();
    });

    it('throws when all partials are disabled for a theme', () => {
      expect(() =>
        getFigmaTheme(createFileObject(brokenFilesIncorectTokens), 'theme4')
      ).toThrowErrorMatchingSnapshot();
    });
  });
});

describe('figmaTheme/getFigmaThemeNames', () => {
  it('gets figma theme names from a files object', () => {
    const result = getFigmaThemeNames(createFileObject(correctFiles));
    expect(result).toMatchSnapshot();
  });

  describe('errors', () => {
    it('throws when a $theme file cannot be found', () => {
      expect(() =>
        getFigmaThemeNames(createFileObject(brokenFilesNoTheme))
      ).toThrowErrorMatchingSnapshot();
    });
  });
});
