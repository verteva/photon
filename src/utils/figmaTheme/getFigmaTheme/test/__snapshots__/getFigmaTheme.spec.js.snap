// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`figmaTheme/getFigmaTheme errors throws when a $meta file cannot be found 1`] = `"No $meta file found"`;

exports[`figmaTheme/getFigmaTheme errors throws when a $theme file cannot be found 1`] = `"No $theme file found"`;

exports[`figmaTheme/getFigmaTheme errors throws when all partials are disabled for a theme 1`] = `"No enabled partials found for this theme theme4. Check the $theme file for this theme."`;

exports[`figmaTheme/getFigmaTheme errors throws when there is no theme data for supplied theme 1`] = `"No data found for theme theme4. This theme may not exist."`;

exports[`figmaTheme/getFigmaTheme gets a figma theme from a files object 1`] = `
Object {
  "colors": Object {
    "black": Object {
      "type": "color",
      "value": "#000000",
    },
    "brand": Object {
      "green": Object {
        "type": "color",
        "value": "#219653",
      },
      "orange": Object {
        "type": "color",
        "value": "orange",
      },
      "purple": Object {
        "type": "color",
        "value": "#6f53f9",
      },
      "red": Object {
        "type": "color",
        "value": "#eb5757",
      },
      "teal": Object {
        "type": "color",
        "value": "#31839b",
      },
    },
    "grey": Object {
      "00": Object {
        "type": "color",
        "value": "#07131f",
      },
      "01": Object {
        "type": "color",
        "value": "#6a6e80",
      },
      "02": Object {
        "type": "color",
        "value": "#949494",
      },
      "03": Object {
        "type": "color",
        "value": "#b5b5b5",
      },
      "08": Object {
        "type": "color",
        "value": "#f4f4f4",
      },
      "09": Object {
        "type": "color",
        "value": "grey",
      },
      "10": Object {
        "type": "color",
        "value": "#f9f9f9",
      },
    },
    "theme": Object {
      "primary": Object {
        "type": "color",
        "value": "{colors.brand.grey}",
      },
      "secondary": Object {
        "type": "color",
        "value": "{colors.brand.teal}",
      },
      "tertiary": Object {
        "type": "color",
        "value": "{colors.brand.red}",
      },
    },
    "white": Object {
      "type": "color",
      "value": "#ffffff",
    },
  },
}
`;

exports[`figmaTheme/getFigmaTheme gets a figma theme from a files object 2`] = `
Object {
  "colors": Object {
    "black": Object {
      "type": "color",
      "value": "#000000",
    },
    "brand": Object {
      "green": Object {
        "type": "color",
        "value": "#219653",
      },
      "orange": Object {
        "type": "color",
        "value": "orange",
      },
      "purple": Object {
        "type": "color",
        "value": "#6f53f9",
      },
      "red": Object {
        "type": "color",
        "value": "#eb5757",
      },
      "teal": Object {
        "type": "color",
        "value": "#31839b",
      },
    },
    "grey": Object {
      "00": Object {
        "type": "color",
        "value": "#07131f",
      },
      "01": Object {
        "type": "color",
        "value": "#6a6e80",
      },
      "02": Object {
        "type": "color",
        "value": "#949494",
      },
      "03": Object {
        "type": "color",
        "value": "#b5b5b5",
      },
      "08": Object {
        "type": "color",
        "value": "#f4f4f4",
      },
      "09": Object {
        "type": "color",
        "value": "grey",
      },
      "10": Object {
        "type": "color",
        "value": "#f9f9f9",
      },
    },
    "theme": Object {
      "primary": Object {
        "type": "color",
        "value": "{colors.brand.grey}",
      },
      "secondary": Object {
        "type": "color",
        "value": "{colors.brand.teal}",
      },
      "tertiary": Object {
        "type": "color",
        "value": "{colors.brand.red}",
      },
    },
    "white": Object {
      "type": "color",
      "value": "#ffffff",
    },
  },
}
`;

exports[`figmaTheme/getFigmaTheme gets a figma theme from a files object with folders 1`] = `
Object {
  "colors": Object {
    "black": Object {
      "type": "color",
      "value": "#000000",
    },
    "brand": Object {
      "green": Object {
        "type": "color",
        "value": "#219653",
      },
      "orange": Object {
        "type": "color",
        "value": "orange",
      },
      "purple": Object {
        "type": "color",
        "value": "#6f53f9",
      },
      "red": Object {
        "type": "color",
        "value": "#eb5757",
      },
      "teal": Object {
        "type": "color",
        "value": "#31839b",
      },
    },
    "grey": Object {
      "00": Object {
        "type": "color",
        "value": "#07131f",
      },
      "01": Object {
        "type": "color",
        "value": "#6a6e80",
      },
      "02": Object {
        "type": "color",
        "value": "#949494",
      },
      "03": Object {
        "type": "color",
        "value": "#b5b5b5",
      },
      "08": Object {
        "type": "color",
        "value": "#f4f4f4",
      },
      "09": Object {
        "type": "color",
        "value": "grey",
      },
      "10": Object {
        "type": "color",
        "value": "#f9f9f9",
      },
    },
    "theme": Object {
      "primary": Object {
        "type": "color",
        "value": "{colors.brand.grey}",
      },
      "secondary": Object {
        "type": "color",
        "value": "{colors.brand.teal}",
      },
      "tertiary": Object {
        "type": "color",
        "value": "{colors.brand.red}",
      },
    },
    "white": Object {
      "type": "color",
      "value": "#ffffff",
    },
  },
  "spacing": Object {
    "default": Object {
      "type": "spacing",
      "value": "2",
    },
  },
}
`;

exports[`figmaTheme/getFigmaTheme gets a figma theme from a files object with folders 2`] = `
Object {
  "colors": Object {
    "black": Object {
      "type": "color",
      "value": "#000000",
    },
    "brand": Object {
      "green": Object {
        "type": "color",
        "value": "#219653",
      },
      "orange": Object {
        "type": "color",
        "value": "orange",
      },
      "purple": Object {
        "type": "color",
        "value": "#6f53f9",
      },
      "red": Object {
        "type": "color",
        "value": "#eb5757",
      },
      "teal": Object {
        "type": "color",
        "value": "#31839b",
      },
    },
    "grey": Object {
      "00": Object {
        "type": "color",
        "value": "#07131f",
      },
      "01": Object {
        "type": "color",
        "value": "#6a6e80",
      },
      "02": Object {
        "type": "color",
        "value": "#949494",
      },
      "03": Object {
        "type": "color",
        "value": "#b5b5b5",
      },
      "08": Object {
        "type": "color",
        "value": "#f4f4f4",
      },
      "09": Object {
        "type": "color",
        "value": "grey",
      },
      "10": Object {
        "type": "color",
        "value": "#f9f9f9",
      },
    },
    "theme": Object {
      "primary": Object {
        "type": "color",
        "value": "{colors.brand.grey}",
      },
      "secondary": Object {
        "type": "color",
        "value": "{colors.brand.teal}",
      },
      "tertiary": Object {
        "type": "color",
        "value": "{colors.brand.red}",
      },
    },
    "white": Object {
      "type": "color",
      "value": "#ffffff",
    },
  },
  "spacing": Object {
    "default": Object {
      "type": "spacing",
      "value": "2",
    },
  },
}
`;

exports[`figmaTheme/getFigmaTheme gets a second figma theme from a files object 1`] = `
Object {
  "colors": Object {
    "black": Object {
      "type": "color",
      "value": "#000000",
    },
    "brand": Object {
      "green": Object {
        "type": "color",
        "value": "#219653",
      },
      "orange": Object {
        "type": "color",
        "value": "purple",
      },
      "purple": Object {
        "type": "color",
        "value": "#6f53f9",
      },
      "red": Object {
        "type": "color",
        "value": "#eb5757",
      },
      "teal": Object {
        "type": "color",
        "value": "#31839b",
      },
    },
    "grey": Object {
      "00": Object {
        "type": "color",
        "value": "#07131f",
      },
      "01": Object {
        "type": "color",
        "value": "#6a6e80",
      },
      "02": Object {
        "type": "color",
        "value": "#949494",
      },
      "03": Object {
        "type": "color",
        "value": "#b5b5b5",
      },
      "08": Object {
        "type": "color",
        "value": "#f4f4f4",
      },
      "09": Object {
        "type": "color",
        "value": "#f6f6f6",
      },
      "10": Object {
        "type": "color",
        "value": "#f9f9f9",
      },
    },
    "theme": Object {
      "primary": Object {
        "type": "color",
        "value": "{colors.brand.black}",
      },
      "secondary": Object {
        "type": "color",
        "value": "{colors.brand.teal}",
      },
      "tertiary": Object {
        "type": "color",
        "value": "{colors.brand.red}",
      },
    },
    "white": Object {
      "type": "color",
      "value": "#ffffff",
    },
  },
}
`;

exports[`figmaTheme/getFigmaTheme gets a second figma theme from a files object 2`] = `
Object {
  "colors": Object {
    "black": Object {
      "type": "color",
      "value": "#000000",
    },
    "brand": Object {
      "green": Object {
        "type": "color",
        "value": "#219653",
      },
      "orange": Object {
        "type": "color",
        "value": "purple",
      },
      "purple": Object {
        "type": "color",
        "value": "#6f53f9",
      },
      "red": Object {
        "type": "color",
        "value": "#eb5757",
      },
      "teal": Object {
        "type": "color",
        "value": "#31839b",
      },
    },
    "grey": Object {
      "00": Object {
        "type": "color",
        "value": "#07131f",
      },
      "01": Object {
        "type": "color",
        "value": "#6a6e80",
      },
      "02": Object {
        "type": "color",
        "value": "#949494",
      },
      "03": Object {
        "type": "color",
        "value": "#b5b5b5",
      },
      "08": Object {
        "type": "color",
        "value": "#f4f4f4",
      },
      "09": Object {
        "type": "color",
        "value": "#f6f6f6",
      },
      "10": Object {
        "type": "color",
        "value": "#f9f9f9",
      },
    },
    "theme": Object {
      "primary": Object {
        "type": "color",
        "value": "{colors.brand.black}",
      },
      "secondary": Object {
        "type": "color",
        "value": "{colors.brand.teal}",
      },
      "tertiary": Object {
        "type": "color",
        "value": "{colors.brand.red}",
      },
    },
    "white": Object {
      "type": "color",
      "value": "#ffffff",
    },
  },
}
`;

exports[`figmaTheme/getFigmaThemeNames errors throws when a $theme file cannot be found 1`] = `"No $theme file found"`;

exports[`figmaTheme/getFigmaThemeNames gets figma theme names from a files object 1`] = `
Array [
  "theme1",
  "theme2",
]
`;
